File: src/main/java/edu/berkeley/veloxms/storage/ModelStorage.java
package edu.berkeley.veloxms.storage;

/**
 * Simple interface to abstract out the KV storage backend used to store
 * the models from the application logic to access them.
 */
public interface ModelStorage {

    /**
     * Get factor vector for provided item
     * @param itemId the item id to get
     */
    public double[] getItemFactors(long itemId);

    public double[] getUserFactors(long userId);

    /**
     * Gets a list of all movies this user has rated and the associated ratings.
     */
    public HashMap<Long, Integer> getRatedMovies (long userId);

}

File: veloxms-core/src/main/scala/edu/berkeley/veloxms/storage/ModelStorage.scala
package edu.berkeley.veloxms.storage

import io.dropwizard.lifecycle.Managed
// import org.codehaus.jackson.JsonNode
// import com.fasterxml.jackson.databind.JsonNode

import scala.util.Try
import edu.berkeley.veloxms._
import scala.collection.JavaConversions._

// import scala.collection.immutable.HashMap

/**
 * Simple interface to abstract out the KV storage backend used to store
 * the models from the application logic to access them.
 *
 * @tparam K The type of the stored keys
 * @tparam V The type of the stored values
 */
trait ModelStorage[K, V] extends Managed {

    override def start() { }

    def put(kv: (K, V)): Unit

    def get(key: K): Option[V]

    // TODO return type?
    def getEntries: Map[K, V] = {
        throw new NotImplementedError("getEntries() not implemented for this storage backend")

    }

}


